bpo-26228: pty.spawn hangs on FreeBSD, OS X, and Solaris

diekmann opened the pull request:
issue26228 as github PR.

This PR contains:
 * original patch of issue26228 by Chris
 * Update of docs
 * Update of pty test suite

According to the bpo discussion, this fixes `pty.spawn()` on FreeBSD, OS X, and Solaris.

<!-- issue-number: bpo-26228 -->
https://bugs.python.org/issue26228
<!-- /issue-number -->

diekmann committed:
original patch by Chris Torek

https://bugs.python.org/file41738/pty.patch

diekmann committed:
Updating documentation for pty module

* Linux -> Unix
* More details about the behavior of spawn

diekmann committed:
Updated pty test suite

diekmann committed:
Acks

vadmium commented on Lib/pty.py:
```diff
@@ -153,7 +158,15 @@ def spawn(argv, master_read=_read, stdin_read=_read):
+            # If we wanted to be really clever, we would use
+            # the same method as subprocess() to pass the error
+            # back to the parent.  For now just dump stack trace.
+            traceback.print_exc()
```
Missing module again. In PR #2932 you fixed this line. See <https://bugs.python.org/review/29070/diff/19699/Lib/pty.py#newcode170>.

diekmann replied:
Thanks for catching this ... again :jack_o_lantern: :ghost: 

Thank you so much for your feedback! :1st_place_medal:

diekmann committed:
Fix missing import in pty.py patch (error path)

Thanks to vadmium for catching this error twice.
https://bugs.python.org/review/29070/diff/19699/Lib/pty.py#newcode170

diekmann committed:
Adding Misc/NEWS.d entry

vadmium commented on Lib/test/test_pty.py:
```diff
@@ -243,67 +254,164 @@ def _pipe(self):
+        return rfds_result, [], []
+
+    def test__mock_stdin_stdout(self):
+        self.assertGreater(pty.STDIN_FILENO, 2, "replaced by our mock")
```
Why is the method called _stdin_stdout_, while the body only looks at STDIN_FILENO?

diekmann replied:
Fixed in the following commit: A docstring explains the name and the method also checks for STDOUT_FILENO.

vadmium commented on Lib/test/test_pty.py:
```diff
@@ -243,67 +254,164 @@ def _pipe(self):
+            # called with three empty lists as file descriptors to wait
+            # on.  Behavior of real select is platform-dependent and
+            # likely infinite blocking on Linux.
+            raise self.fail("mock select on no waitables")
```
I suspect the _raise_ statement never executes because _TestCase.fail_ raises its own exception.

diekmann replied:
Your [suspicion](https://github.com/python/cpython/blob/032a6480e360427d4f964e31643604fad804ea14/Lib/unittest/case.py#L678) is appropriate. Fixed in the following commit.

vadmium commented on Lib/test/test_pty.py:
```diff
@@ -243,67 +254,164 @@ def _pipe(self):
+
+        # Test that STDIN was not touched.  This test simulated the
+        # scenario where the child process immediately closed its end of
+        # the pipe.  This means, nothing should be copied.
```
What is the point of this test? I don’t think it matters.

You seem to be modelling a situation where input (or EOF) to the child becomes available at about the same instant that the slave terminal is closed. If __copy_ were written differently, it might copy the input to the child (STDIN_FILENO) _before_ checking the pseudoterminal master (_master_fd_), and the test would fail.

diekmann replied:
True, this part of the test is overly fitted for an unnecessary implementation detail. Removed in the following commit

vadmium reviewed (commented):
I can’t see anything seriously wrong with 005dc23, although I think it would be better to test the API (_pty.spawn_), without depending on the internals (__copy_, _select_, etc) if possible.

diekmann committed:
Incorporate vadmium's suggestions.
